### PR TITLE
Implement AddEowPhonemesToLexiconJob

### PR DESCRIPTION
**Pull request by SimBe195**
Implements new `AddEowPhonemesToLexiconJob` which adds eow-phonemes to a lexicon. Such an extended lexicon is useful for recognition with the upcoming GenericSeq2Seq decoder.

### CODE REVIEW

File: lexicon/modification.py [hunk 1]
```diff
@@ -185,3 +186,56 @@ def run(self):
...
+
+    def run(self):
+        in_lex = lexicon.Lexicon()
+        in_lex.load(self.bliss_lexicon.get_path())
+
+        out_lex = lexicon.Lexicon()
+
+        # Identify all 'special' phonemes in the given lexicon. A phoneme is
+        # deemed 'special' if it appears in the pronunciation of a special lemma.
+        # This should identify e.g. 'silence', 'blank', 'unknown' etc.
```
Comment (michelwi) [inline lexicon/modification.py:215]: how about `[noise]` and similar?

File: lexicon/modification.py [hunk 2]
```diff
@@ -185,3 +186,56 @@ def run(self):
...
+                continue
+            for phon in lemma.phon:
+                special_phonemes.update(phon.split())
+
+        # Add all phonemes to out_lexicon and create list of eow-phonemes
+        eow_phonemes = []
+        for phoneme, variation in in_lex.phonemes.items():
+            out_lex.add_phoneme(phoneme, variation)
+            if phoneme not in special_phonemes:
+                eow_phonemes.append((phoneme + "#", variation))
```
Comment (michelwi) [inline lexicon/modification.py:228]: would it make sense to make the EOW marker `"#"` configurable by the user?

Comment (SimBe195) [reply lexicon/modification.py:228]: Sure, why not.

File: lexicon/modification.py (hunk 1 above)
Comment (SimBe195) [reply lexicon/modification.py:215]: Yes, fixed now by adding a `nonword_phones` parameter to the job where such phonemes can be specified.

File: lexicon/modification.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
-__all__ = ["WriteLexiconJob", "MergeLexiconJob"]
+__all__ = ["WriteLexiconJob", "MergeLexiconJob", "AddBoundaryMarkerToLexiconJob"]
```
Comment (JackTemaki) [inline lexicon/modification.py:1]: Does not match the job name

**Review by JackTemaki — CHANGES_REQUESTED** [review]

File: lexicon/modification.py [hunk 4]
```diff
@@ -185,3 +186,63 @@ def run(self):
...
+            if phoneme not in special_phonemes:
+                eow_phonemes.append((phoneme + self.boundary_marker, variation))
+
+        # Add eow-phonemes to out_lexicon
+        for eow_phoneme, variation in eow_phonemes:
+            out_lex.add_phoneme(eow_phoneme, variation)
+
+        # Modify lemmata to include eow-phoneme
+        for lemma in in_lex.lemmata:
+            if lemma.special is None and lemma.phon is not None:
```
Comment (michelwi) [inline lexicon/modification.py:243]: And you also need to check, that the (last?) phone of the lemma is not in `special_phones` (or more precise `self.nonword_phones`) as here I believe you add the boundary marker to the noise lemma.

Comment (SimBe195) [reply lexicon/modification.py]: Yes that's true. Fixed now.

File: lexicon/modification.py [hunk 5]
```diff
@@ -210,6 +210,32 @@ def __init__(self, bliss_lexicon: tk.Path, nonword_phones: Optional[List] = None
...
+        :param pron: Original pronunciation as a string containing phonemes separated by whitespaces
+        :param special_phonemes: Set of special phonemes that should be skipped over.
+        :return: Modified pronunciation
+        """
+        phoneme_list = pronunciation.split()
+
+        for i, phoneme in reversed(list(enumerate(phoneme_list))):
+            if phoneme in special_phonemes:
+                continue
+            phoneme_list[i] = self._eow_phoneme(phoneme)
```
Comment (michelwi) [inline lexicon/modification.py:235]:
```suggestion
            phoneme_list[i] = self._eow_phoneme(phoneme)
            break
```
I think you are currently adding eow marker to all non-special phonemes. 
Fix this by ending the for-loop after first match.

File: lexicon/modification.py [hunk 6]
```diff
@@ -185,3 +186,89 @@ def run(self):
...
+        return phoneme + self.boundary_marker
+
+    def _modify_pronunciation(self, pronunciation: str, special_phonemes: Set[str]) -> str:
+        """
+        Find the rightmost phoneme in the pronunciation that is not special and replace it by its
+        EOW version. Might do nothing if all phonemes are special.
+
+        Example: "AA BB [noise]" -> "AA BB# [noise]"
+
+        :param pron: Original pronunciation as a string containing phonemes separated by whitespaces
```
Comment (Atticus1806) [inline lexicon/modification.py:226]:
```suggestion
        :param pronunciation: Original pronunciation as a string containing phonemes separated by whitespaces
```

File: lexicon/modification.py [hunk 7]
```diff
@@ -185,3 +186,89 @@ def run(self):
...
+        self.boundary_marker = boundary_marker
+
+        self.out_lexicon = self.output_path("lexicon.xml")
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def _eow_phoneme(self, phoneme: str) -> str:
+        """
+        Creates the EOW version of a given phoneme.
```
Comment (Atticus1806) [inline lexicon/modification.py:215]:
```suggestion
        Creates the eow-version of a given phoneme.
```

**Review by Atticus1806 — APPROVED** [review]
Looks good. Only minor doc stuff.

File: lexicon/modification.py [hunk 8]
```diff
@@ -185,3 +186,89 @@ def run(self):
...
+        """
+        return phoneme + self.boundary_marker
+
+    def _modify_pronunciation(self, pronunciation: str, special_phonemes: Set[str]) -> str:
+        """
+        Find the rightmost phoneme in the pronunciation that is not special and replace it by its
+        EOW version. Might do nothing if all phonemes are special.
```
Comment (christophmluscher) [inline lexicon/modification.py:222]:
```suggestion
        eow-version. Might do nothing if all phonemes are special.
```

**Review by JackTemaki — APPROVED** [review]

**Pull request merged**